### PR TITLE
Unblock NumPy 2.0

### PR DESCRIPTION
**Pull request by NeilGirdhar**
Fixes https://github.com/huggingface/datasets/issues/6980

### CODE REVIEW

**Comment by NeilGirdhar** [comment]
@albertvillanova Any chance we could get this in before the next release?  Everything depending on HuggingFace has their NumPy upgrade blocked.

**Review by albertvillanova — COMMENTED** [review]
The CI integration tests for Python 3.10 are failing.

**Review by albertvillanova — CHANGES_REQUESTED** [review]
Note that the CI tests for Python 3.8 are OK because they use numpy 1.24.4: numpy 2.0.0 requires Python >= 3.10

**Comment by albertvillanova** [comment]
The incompatible libraries are:
- faiss-cpu 1.8.0.post1 requires numpy<2.0,>=1.0, but you have numpy 2.0.0 which is incompatible.
- tensorflow 2.16.2 requires numpy<2.0.0,>=1.23.5; python_version <= "3.11", but you have numpy 2.0.0 which is incompatible.
- transformers 4.42.3 requires numpy<2.0,>=1.17, but you have numpy 2.0.0 which is incompatible.

**Comment by NeilGirdhar** [comment]
Why is it installing numpy 2 if the dependencies don't support it?

**Comment by NeilGirdhar** [comment]
For me, I'm getting:
```
❯ uv pip install --system "datasets[tests] @ ."
Found existing alias for "uv pip install". You should use: "pipi"
Resolved 119 packages in 934ms
   Built datasets @ file:///Users/neil/src/datasets
Prepared 1 package in 1.28s
Uninstalled 1 package in 10ms
Installed 2 packages in 17ms
 - datasets==2.20.1.dev0 (from file:///Users/neil/src/datasets)
 + datasets==2.20.1.dev0 (from file:///Users/neil/src/datasets)
 + numpy==1.26.4
```

**Comment by albertvillanova** [comment]
Which version on Python do you have?

**Comment by NeilGirdhar** [comment]
3.12.4  I'll try on 3.10 now.

**Comment by albertvillanova** [comment]
Please, note that I obtained  the previous incompatible libraries in my local environment, by forcing the update of numpy.

**Comment by albertvillanova** [comment]
In the Python 3.10 CI, the situation is different:
- for example, they install an older version of tensorflow (2.14.0), where probably the constraint on numpy was not yet implemented. See the details: https://github.com/huggingface/datasets/actions/runs/9879100332/job/27306903343?pr=6991
```
> uv pip install --system "datasets[tests] @ ."
...
 + faiss-cpu==1.8.0
...
  + numpy==2.0.0
...
 + tensorflow==2.14.0
```

**Comment by albertvillanova** [comment]
The main point is that we cannot support numpy 2.0 until tensorflow and faiss do.

**Comment by NeilGirdhar** [comment]
> Alternatively, we should ignore/select tests depending on the installed versions.

That works.

Alternatively, you could depend on tensorflow 2.16.2 (etc.) for the tests?

**Comment by albertvillanova** [comment]
Yes, I was thinking of a workaround solution.

The issue I see is that our CI will not test numpy 2.0 indeed.

**Comment by NeilGirdhar** [comment]
> The issue I see is that our CI will not test numpy 2.0 indeed.

Right, that's the advantage of the test skipping you wanted, I see your point.

**Comment by albertvillanova** [comment]
Without testing Numpy 2.0, we do not know if there are some other parts in the code broken.

**Comment by albertvillanova** [comment]
> I guess the solution is to write tests that don't depend on tensorflow, etc., but still use numpy.
That is my point. What we cannot do is just blindly support Numpy 2.0 without knowing its consequences. We need to test it:
- to know if our core code works with it
- to know what optional libraries are incompatible

For example, while testing locally, I have discovered that librosa is also incompatible with numpy-2.0, due to its dependency on soxr...

**Comment by albertvillanova** [comment]
While testing locally, I have also discovered that pytorch does not support Numpy 2.0 on Windows platforms.

**Comment by albertvillanova** [comment]
I am adding Numpy 2.0 tests to your PR if you don't mind, before merging this PR.

**Comment by NeilGirdhar** [comment]
Awesome, thank you!  Please let me know if I need to do anything.

**Comment by albertvillanova** [comment]
Now we test numpy 2.0 in the `test_py310_numpy2` CI tests: https://github.com/huggingface/datasets/actions/runs/9907254874/job/27370545495?pr=6991
```
 + numpy==2.0.0
```

**Review by albertvillanova — APPROVED** [review]
Thank you.
